Finish Introdution To Testing (#519)
campuscode/rails-guides-pt-BR: pt-BR/testing.md

Os testes no Rails podem simular requisições no browser e com isso, você pode testar a resposta da sua aplicação sem ter que testar utilizando de fato seu o browser.

Introdução a testes
-----------------------

> Suporte a testes foi implantado no Rails desde os primórdios. Não foi uma epifania do tipo: "Vamos adicionar suporte para testes porque eles são novos e legais!"

Reviewer: ```suggestion
O suporte a testes foi implantado no Rails desde os primórdios. Não foi uma epifania do tipo: "Vamos adicionar suporte para testes porque eles são novos e legais!"
```

### Configurações para testes em aplicações Rails

> Rails cria um diretório `test` para você logo que você cria um projeto Rails usando o comando `rails new` _nome_da_aplicacao_. Se você deseja listar o conteúdo deste diretório, então você vai ver:

Reviewer: ```suggestion
O Rails cria um diretório `test` para você logo que você cria um projeto Rails usando o comando `rails new` _nome_da_aplicacao_. Se você listar o conteúdo deste diretório, você verá:
```

```bash
$ ls -F test
application_system_test_case.rb  controllers/                     helpers/                         mailers/                         system/
channels/                        fixtures/                        integration/                     models/                          test_helper.rb
```

> Os diretórios `helpers`, `mailers` e `models` são destinados a realizar os testes para *view helpers*, *mailers* e *models*, respectivamente. O diretório `channel` é destinado a realizar os testes para a conexão e canais do *ActionCable*. O diretório `controllers` se destina a realizar testes para os *controllers*, rotas e *views*. O diretório `integration` se destina a realizar testes entre a interação entre *controllers*.

Reviewer: ```suggestion
Os diretórios `helpers`, `mailers` e `models` são destinados a realizar os testes para *view helpers*, *mailers* e *models*, respectivamente. O diretório `channel` é destinado a realizar os testes para a conexão e canais do *ActionCable*. O diretório `controllers` se destina a realizar testes para os *controllers*, rotas e *views*. O diretório `integration` se destina a realizar testes de interação entre *controllers*.
```

> O diretório `system` é destinado a realizar os testes do sistema, que são usados para testes completos da aplicação no browser. Testes de Sistema permitem você testar a aplicação do jeito que seu usuário experiência e ajuda você a testar seu JavaScript também.

Reviewer: ```suggestion
O diretório `system` é destinado a realizar os testes do sistema, que são usados para testes completos da aplicação no browser. Os Testes de Sistema permitem você testar a aplicação do jeito que seu usuário experiência e ajuda você a testar seu JavaScript também.
```

> Testes de sistemas herdam de Capybara e performam os testes de *browser* para a sua aplicação

Reviewer: ```suggestion
Os Testes de sistemas herdam de Capybara e são executados em testes de *browser* para a sua aplicação
```

*Fixtures* são um jeito de organizar dados de testes; ficam no diretório `fixtures`

Um diretório `jobs` também será criado quando um teste associado é gerado.

O arquivo `test_helper.rb` é responsável por realizar as configurações padrão para seus testes.

O arquivo `application_system_test_case.rb` é responsável por realizar as configurações padrão para seus testes de sistema.

### O Ambiente de Teste

Por padrão, toda aplicação Rails tem três ambientes: desenvolvimento, teste e produção.

> Cada configuração de ambiente pode ser modificada similarmente. Neste caso, nós podemos modificar nosso ambiente de teste por alterar as opções encontradas em `config/environments/test.rb`.

Reviewer: ```suggestion
A configuração de cada ambiente pode ser modificada de forma semelhante. Neste caso, podemos modificar nosso ambiente de teste alterando as opções encontradas em `config/environments/test.rb`.
```

NOTE: Seus testes são executados sob o comando `RAILS_ENV=test`.

### Rails conhece Minitest

> Se você se lembra, nós usamos o comando `bin/rails generate model` no guia [Começando com Rails](getting_started.html). Nós criamos nosso primeiro model, e entre outras coisas foi criado rascunho de testes no diretório `test`:

Reviewer: ```suggestion
Se você se lembra, nós usamos o comando `bin/rails generate model` no guia [Começando com Rails](getting_started.html). Nós criamos nosso primeiro model e, entre outras coisas, foi criado [_stub_](https://pt.wikipedia.org/wiki/Stub) de testes no diretório `test`:
```

Acho que não traduziria o _stub_ aqui, pois é como nos referimos no dia a dia. O que acham de deixar o _stub_ e deixar um link de referência para consulta? @vinibispo @campuscode/guia-rails-team

Os testes no Rails podem simular requisições no browser e com isso, você pode testar a resposta da sua aplicação sem ter que testar utilizando de fato seu o browser.

Introdução a testes
-----------------------

Suporte a testes foi implantado no Rails desde os primórdios. Não foi uma epifania do tipo: "Vamos adicionar suporte para testes porque eles são novos e legais!"

### Configurações para testes em aplicações Rails

Rails cria um diretório `test` para você logo que você cria um projeto Rails usando o comando `rails new` _nome_da_aplicacao_. Se você deseja listar o conteúdo deste diretório, então você vai ver:

```bash
$ ls -F test
application_system_test_case.rb  controllers/                     helpers/                         mailers/                         system/
channels/                        fixtures/                        integration/                     models/                          test_helper.rb
```

> Os diretórios `helpers`, `mailers` e `models` são destinados a realizar os testes para view helpers, mailers e models, respectivamente. O diretório `channel` é destinado a realizar os testes para a conexão e canais do ActionCable. O diretório `controllers` se destina a realizar testes para os controllers, rotas e views. O diretório `integration` se destina a realizar testes entre a interação entre controllers.

Reviewer 2: ```suggestion
Os diretórios `helpers`, `mailers` e `models` são destinados a realizar os testes para *view helpers*, *mailers* e *models*, respectivamente. O diretório `channel` é destinado a realizar os testes para a conexão e canais do *ActionCable*. O diretório `controllers` se destina a realizar testes para os *controllers*, rotas e *views*. O diretório `integration` se destina a realizar testes entre a interação entre *controllers*.
```

Acho que aqui seria legal ter alguns itálicos nos estrangerismos o que acha?

Author: Excelente sugestão

O diretório `system` é destinado a realizar os testes do sistema, que são usados para testes completos da aplicação no browser. Testes de Sistema permitem você testar a aplicação do jeito que seu usuário experiencia e ajuda você a testar seu JavaScript também.

> Testes de sistemas herdam de Capybara e performam os testes do browser para a sua aplicação

Author: de browser acredito eu

Fixtures são um jeito de organizar dados de testes; ficam no diretório `fixtures`

> 

Author: Boa kskskksks

Um diretório `jobs` também será criado quando um teste associado é gerado.

> O `test_helper.rb` é responsável por realizar as configurações padrão para seus testes.

Reviewer 2: Acho que faltou uma palavra, pode conferir?

O `application_system_test_case.rb` é responsável por realizar as configurações padrão para seus testes de sistema.

### O Ambiente de Teste

> Por padrão, toda aplicação Rals tem três ambientes: desenvolvimento, teste e produção.

Reviewer 2: ```suggestion
Por padrão, toda aplicação Rails tem três ambientes: desenvolvimento, teste e produção.
```

Author: Minha nossa, como esqueci do Rails ksksksks

...
```

> O teste padrão esboçado em `test/models/article_test.rb` parece assim:

Reviewer: ```suggestion
O _stub_ teste padrão em `test/models/article_test.rb` parece assim:
```

end
```

Uma inspeção linha a linha desse arquivo ajudará você a se orientar a terminologia e código de testes no Rails.

```ruby
require "test_helper"
```

> Por fazer *require* desse arquivo, `test_helper.rb` as configurações padrão para executar nossos testes são carregadas. Nós vamos incluir isso em todos os testes que escrevermos, então qualquer método adicionado a este arquivo vão estar disponíveis em todos os nossos testes.

Reviewer: ```suggestion
Por fazer *require* desse arquivo, `test_helper.rb` as configurações padrões para executar nossos testes são carregadas. Nós vamos incluir isso em todos os testes que escrevermos, então qualquer método adicionado a este arquivo vai estar disponível em todos os nossos testes.
```

```ruby
class ArticleTest < ActiveSupport::TestCase
```

A classe `ArticleTest` define um _test case_ porque ela herda de `ActiveSupport::TestCase`. `ArticleTest` portanto tem todos os métodos disponíveis de `ActiveSupport::TestCase`. Mais pra frente nesse guia, nós vamos adicionar alguns dos métodos dados.

Qualquer método definido com uma classe herdada de `Minitest::Test` (que é uma superclasse de `ActiveSupport::TestCase`) que começa com `test_` é simplesmente chamada em um teste. Então, métodos definidos como `test_password` e `test_valid_password` são nomes de testes legais e serão executados automaticamente quando o caso de teste (_test_case_) é executado.

> Rails também adiciona um método `test` que leva um nome _test_ e um bloco. Isso gera um teste normal `Minitest::Unit` com nomes de métodos prefixados com `test_`. Então você não precisa se preocupar com nomear os métodos, e você pode escrever algo assim:

Reviewer: ```suggestion
O Rails também adiciona um método `test` que leva um nome _test_ e um bloco. Isso gera um teste normal `Minitest::Unit` com nomes de métodos prefixados com `test_`. Então você não precisa se preocupar com nomear os métodos, e você pode escrever algo assim:
```

```ruby
test "the truth" do
  assert true
end
```

Que é aproximadamente o mesmo que escrever isto:

```ruby
def test_the_truth
  assert true
end
```

Apesar de você ainda poder usar definições comuns de métodos, usando o prefixo `test` permite você ter um teste mais legível.

> NOTE: O nome do método é gerado por alterar espaços com undescores(*_*). O resultado não precisa ser um identificador Ruby valido embora o nome possa conter caracteres de pontuação, etc. Isso é porque em Ruby tecnicamente qualquer string pode ser o nome de um método. Isso pode requerer a chamada dos métodos `define_method` e `send` para funcionar corretamente, mas formalmente há uma pequena restrição no nome.

Reviewer: ```suggestion
NOTE: O nome do método é gerado por alterar espaços com undescores(*_*). O resultado não precisa ser um identificador Ruby válido, embora o nome possa conter caracteres de pontuação, etc. Isso é porque em Ruby tecnicamente qualquer string pode ser o nome de um método. Isso pode requerer a chamada dos métodos `define_method` e `send` para funcionar corretamente, mas formalmente há uma pequena restrição no nome.
```

> Próximo, vamos olhar para nossa primeira asserção:

Reviewer: ```suggestion
A seguir, vamos olhar para nossa primeira asserção:
```

class ArticleTest < ActiveSupport::TestCase
```

A classe `ArticleTest` define um _test case_ porque ela herda de `ActiveSupport::TestCase`. `ArticleTest` portanto tem todos os métodos disponíveis de `ActiveSupport::TestCase`. Mais pra frente nesse guia, nós vamos adicionar alguns dos métodos dados.

Qualquer método definido com uma classe herdada de `Minitest::Test` (que é uma superclasse de `ActiveSupport::TestCase`) que começa com `test_` é simplesmente chamada em um teste. Então, métodos definidos como `test_password` e `test_valid_password` são nomes de testes legais e serão executados automaticamente quando o caso de teste (_test_case_) é executado.

Rails também adiciona um método `test` que leva um nome _test_ e um bloco. Isso gera um teste normal `Minitest::Unit` com nomes de métodos prefixados com `test_`. Então você não precisa se preocupar com nomear os métodos, e você pode escrever algo assim:

```ruby
test "the truth" do
  assert true
end
```

Que é aproximadamente o mesmo que escrever isto:

```ruby
def test_the_truth
  assert true
end
```

Apesar de você ainda poder usar definições comuns de métodos, usando o prefixo `test` permite você ter um teste mais legível.

> NOTE: O nome do método é gerado por alterar espaços com undescores(*_*). O resultado não precisa ser um identificador Ruby valido embora o nome possa conter caracteres de pontuação, etc. Isso é porque em Ruby tecnicamente qualquer string pode ser o nome de um método. Isso pode requerir a chamada dos métodos `define_method` e `send` para funcionar corretamente, mas formalmente há uma pequena restrição no nome.

Reviewer 2: ```suggestion
NOTE: O nome do método é gerado por alterar espaços com undescores(*_*). O resultado não precisa ser um identificador Ruby valido embora o nome possa conter caracteres de pontuação, etc. Isso é porque em Ruby tecnicamente qualquer string pode ser o nome de um método. Isso pode requerer a chamada dos métodos `define_method` e `send` para funcionar corretamente, mas formalmente há uma pequena restrição no nome.
```

Próximo, vamos olhar para nossa primeira asserção:

```ruby
assert true
```

Uma asserção é uma linha de código que pode se tornar um objeto (ou expressão) para resultados esperados. Por exemplo, uma asserção pode checar se:

* Esse valor é igual a aquele valor?
* Esse objeto é nulo?
* Essa linha de código dispara uma exceção?
* A senha do usuário é maior que 5 caracteres?

Todo teste pode conter uma ou mais asserções, sem restrições em quantas asserções são permitidas. Apenas quando todas as asserções serem bem-sucedidas o teste vai passar.

> #### Sua primeira falha no teste

Reviewer 2: Acho que essa frase ficou um pouco estranha, o que acha de `Seu primeiro teste de falha`? Será que troca o sentido?

Author: Eu acho q troca, vou pensar numa melhor, o q acha de `Seu primeiro teste que falha`

end
```

> Vamos executar o teste com esse caso de teste adicionado (one 6 é o número da linha onde o teste é definido).

Reviewer: ```suggestion
Vamos executar o teste com esse caso de teste adicionado (onde 6 é o número da linha onde o teste é definido).
```

1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

> Na saída do teste, `F` significa uma falha. Você pode ver o _trace_ correspondendente mostrado abaixo de `Failure` junto com o nome do teste que está falhando. As próximas linhas contém as origens do erro seguido por uma mensagem que menciona o valor atual e o valor esperado pela asserção. A mensagem de asserção padrão provê informação o suficiente para ajudar a localizar o erro. Para fazer a asserção mais legível, toda asserção tem um parâmetro de mensagem opcional, como foi mostrado aqui:

Reviewer 2: ```suggestion
Na saída do teste, `F` significa uma falha. Você pode ver o _trace_ correspondente mostrado abaixo de `Failure` junto com o nome do teste que está falhando. As próximas linhas contém as origens do erro seguido por uma mensagem que menciona o valor atual e o valor esperado pela asserção. A mensagem de asserção padrão provê informação o suficiente para ajudar a localizar o erro. Para fazer a asserção mais legível, toda asserção tem um parâmetro de mensagem opcional, como foi mostrado aqui:
```

Author: Meus erros de pt tão fogo ksksksks

end
```

> Executando esse teste mostra uma mensagem de asserção mais amigável:

Reviewer: ```suggestion
A execução desse teste exibe uma mensagem de asserção mais amigável:
```

end
```

Executando esse teste mostra uma mensagem de asserção mais amigável:

```
Failure:
ArticleTest#test_should_not_save_article_without_title [/path/to/blog/test/models/article_test.rb:6]:
Saved the article without a title
```

> Agora para fazer esse teste pasar nós podemos adicionar uma validação no nível do model para o campo _title_.

Reviewer 2: ```suggestion
Agora para fazer esse teste passar nós podemos adicionar uma validação no nível do model para o campo _title_.
```

1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
```

Agora, se você notou, nós escrevemos um teste que falha para uma funcionalidade desejada, então nós escrevemos um código básico que adiciona a funcionalidade e finalmente nós tivemos certeza que nosso testes passam. Essa abordagem em desenvolvimento de software é referida como
[_Test-Driven Development_ (TDD)](http://c2.com/cgi/wiki?TestDrivenDevelopment).

#### Como um erro se parece

Para ver como um é reportado, aqui está um teste contendo um erro:

```ruby

> test "should report error" do

Reviewer 2: Será que devemos traduzir esse e outros comentários? O que acha?

Author: Boa pergunta, acho que devemos perguntar pro core team

Reviewer 2: @campuscode/guia-rails-team @tkusuki @akaninja @Auralcat @erikacamposdesign 

O que acham?

Reviewer: > @campuscode/guia-rails-team @tkusuki @akaninja @Auralcat @erikacamposdesign
> 
> O que acham?

Eu sou da opinião que se dá pra traduzir (e faz sentido, claro...) acho que sempre vale.
Penso que é uma barreira a menos para o aprendizado.

Reviewer 2: Bora então? Você pode fazer a alteração @vinibispo ?

| `assert_respond_to( obj, symbol, [msg] )`                        | Checa se `obj` responde a `symbol`.|
| `assert_not_respond_to( obj, symbol, [msg] )`                    | Checa se `obj` não responde a `symbol`.|
| `assert_operator( obj1, operator, [obj2], [msg] )`               | Checa se `obj1.operator(obj2)` é verdadeiro.|
| `assert_not_operator( obj1, operator, [obj2], [msg] )`           | Checa se `obj1.operator(obj2)` é falso.|
| `assert_predicate ( obj, predicate, [msg] )`                     | Checa se `obj.predicate` é verdadeiro, ex: `assert_predicate str, :empty?`|
| `assert_not_predicate ( obj, predicate, [msg] )`                 | Checa se `obj.predicate` é falso, ex: `assert_not_predicate str, :empty?`|
| `flunk( [msg] )`                                                 | Checa que o teste falha. Isso é útil para marcar explicitamente que o teste não está finalizado ainda.|

As asserções acima são um subconjunto de asserções que `minitest` suporta. Para uma lista mais atualizada, por favor cheque
[Minitest API documentation](http://docs.seattlerb.org/minitest/), especificamente
[`Minitest::Assertions`](http://docs.seattlerb.org/minitest/Minitest/Assertions.html).

> Por conta da natureza modular de `frameworks` de testes, é possível criar suas próprias asserções. De fato, isso é exatamente o que Rails faz. Isso inclui algumas asserções especializadas para deixar sua vida mais fácil.

Reviewer 2: ```suggestion
Por conta da natureza modular do `framework` de testes, é possível criar suas próprias asserções. De fato, isso é exatamente o que Rails faz. Isso inclui algumas asserções especializadas para deixar sua vida mais fácil.
```

Você acha que aqui eles quiseram dizer no geral ou especificamente do minitest? fiquei nessa dúvida

Author: Acho q eles quiseram dizer do minitest em específico, vou alterar

* [`ActionDispatch::SystemTestCase`](https://api.rubyonrails.org/classes/ActionDispatch/SystemTestCase.html)
* [`Rails::Generators::TestCase`](https://api.rubyonrails.org/classes/Rails/Generators/TestCase.html)

Cada uma dessas asserções inclui `Minitest::Assertions`, nos permitindo usar todas as asserções básicas de nossos testes.

NOTE: Para mais informações em `Minitest`, procure na sua [própria documentação](http://docs.seattlerb.org/minitest).

> ### O `Teste Runner` do rails

Reviewer 2: ```suggestion
### O `Teste Runner` do Rails
```

1 tests, 1 assertions, 0 failures, 0 errors, 0 skips
```

Você pode também executar uma linha específica de um teste colocando o número da linha.

```bash
$ bin/rails test test/models/article_test.rb:6 # run specific test and line
```

Você também pode executar um diretório inteiro de testes colocando o caminho do diretório.

```bash
$ bin/rails test test/controllers # run all tests from specific directory
```

> O Teste Runner provê muitas feature como _failing fast_, adiando a saída do teste
> no fim da execução do teste e assim por diante. Cheque a documentação do Test Runner conferme a seguir:

Reviewer 2: ```suggestion
O Teste Runner provê muitas feature como _failing fast_, adiando a saída do teste
no fim da execução do teste e assim por diante. Cheque a documentação do Test Runner da seguinte forma:
```
